x-pack/solutions/observability/plugins/apm/server/routes/service_map/fetch_service_paths_from_trace_ids.ts

```diff
@@ -176,65 +176,66 @@ export async function fetchServicePathsFromTraceIds({
             }
       
             def processAndReturnEvent(def context, def eventId) {
-              def stack = new Stack();
-              def reprocessQueue = new LinkedList();
-
-              // Avoid reprocessing the same event
+              def pathStack = new Stack();
               def visited = new HashSet();
 
-              stack.push(eventId);
+              def event = context.eventsById.get(eventId);
+
+              if (event == null) {
+                return null;
+              }
+              
+              pathStack.push(eventId);
+
+              // build a stack with the path from the current event to the root
+              def parentId = event['parent.id'];
+              while (parentId != null && !parentId.equals(eventId)) {
+                def parent = context.eventsById.get(parentId);
+                if (parent == null || visited.contains(parentId)) {
+                  break;
+                }
+                  
+                pathStack.push(parentId);
+                visited.add(parentId);
+                parentId = parent['parent.id'];
+              }
+
+              // pop the stack starting from the root to current event to build the path
+              while (!pathStack.isEmpty()) {
```

Comment:
With the path from current event to root transaction defined, it iterates over the stack from the root transaction up until the current event.